providers/azure.go

```diff
@@ -149,13 +157,40 @@ func (p *AzureProvider) Redeem(ctx context.Context, redirectURL, code string) (*
 	created := time.Now()
 	expires := time.Unix(jsonResponse.ExpiresOn, 0)
 
-	return &sessions.SessionState{
+	session := &sessions.SessionState{
 		AccessToken:  jsonResponse.AccessToken,
 		IDToken:      jsonResponse.IDToken,
 		CreatedAt:    &created,
 		ExpiresOn:    &expires,
 		RefreshToken: jsonResponse.RefreshToken,
-	}, nil
+	}
+
+	// the presence of id_token depends on the configuration and
+	// verification of id_token requires oidc configuration
+	if session.IDToken != "" && p.Verifier != nil {
+		email, err := p.verifyIDTokenAndExtractEmail(ctx, session.IDToken)
+		if err != nil {
+			return nil, err
+		}
+		session.Email = email
+	}
+
+	return session, nil
+}
+
+func (p *AzureProvider) verifyIDTokenAndExtractEmail(ctx context.Context, rawIDToken string) (string, error) {
+	idToken, err := p.Verifier.Verify(ctx, rawIDToken)
+	if err != nil {
+		return "", err
+	}
+	var idTokenClaims struct {
+		Email string `json:"email,omitempty"`
+	}
+	if err := idToken.Claims(&idTokenClaims); err != nil {
+		return "", err
+	}
+
+	return idTokenClaims.Email, nil
```

Comment:
Is the `email_verified` claim a thing in Azure? If so, I think we need to respect that claim & related options we have in oauth2-proxy around trusting emails if they aren't verified.  Here's a sample: https://github.com/oauth2-proxy/oauth2-proxy/blob/d2ffef2c7e0a622295c42cb2fdd70ef14593d03c/providers/provider_data.go#L163

Reply:
it's not. https://docs.microsoft.com/en-us/azure/active-directory/develop/active-directory-optional-claims